jax/scipy/ndimage.py

```diff
@@ -0,0 +1,112 @@
+# Copyright 2018 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import functools
+import itertools
+import operator
+import textwrap
+
+import scipy.ndimage
+
+from ..numpy import lax_numpy as np
+from ..numpy.lax_numpy import _wraps
+from ..util import safe_zip as zip
+
+
+_nonempty_prod = functools.partial(functools.reduce, operator.mul)
+_nonempty_sum = functools.partial(functools.reduce, operator.add)
+
+_INDEX_FIXERS = {
+    'constant': lambda index, size: index,
+    'nearest': lambda index, size: np.clip(index, 0, size - 1),
+    'wrap': lambda index, size: index % size,
+}
+
+
+def _nearest_indices_and_weights(coordinate):
+  index = np.around(coordinate).astype(np.int32)
+  weight = coordinate.dtype.type(1)
+  return [(index, weight)]
+
+
+def _linear_indices_and_weights(coordinate):
+  lower = np.floor(coordinate)
+  upper = np.ceil(coordinate)
+  l_index = lower.astype(np.int32)
+  u_index = upper.astype(np.int32)
+  one = coordinate.dtype.type(1)
+  l_weight = one - (coordinate - lower)
+  u_weight = one - l_weight  # handles the edge case lower==upper
+  return [(l_index, l_weight), (u_index, u_weight)]
+
+
+@_wraps(scipy.ndimage.map_coordinates, lax_description=textwrap.dedent("""\
+    Only linear interpolation (``order=1``) and modes ``'constant'``,
+    ``'nearest'`` and ``'wrap'`` are currently supported. Note that
+    interpolation near boundaries differs from the scipy function, because we
+    fixed an outstanding bug (https://github.com/scipy/scipy/issues/2640);
+    this function interprets the ``mode`` argument as documented by SciPy, but
+    not as implemented by SciPy.
+    """))
+def map_coordinates(
+    input, coordinates, order, mode='constant', cval=0.0,
+):
+  input = np.asarray(input)
+  coordinates = [np.asarray(c, input.dtype) for c in coordinates]
```

Comment:
Shouldn't this accept an array, not a list of arrays? It seems much more likely to scale to large coordinate lists that way. It seems highly likely that folks would want to pass thousands of coordinates here.

(If that sounds annoying to do, I have one word, which is `vmap`.)

Reply:
The first dimension of `coordinates` coordinates to `input.ndim`, e.g., for interpolation on a 3D array it is of size 3. So I think this is fine -- the complexity of linear interpolation is `O(2**input.ndim)` anyways, so this dimension is probably never going to be much larger, anyways.

Reply:
I considered adding a comment, but decided to actually check this condition instead.